core/src/main/java/com/scalar/db/util/groupcommit/Group.java

```diff
@@ -0,0 +1,245 @@
+package com.scalar.db.util.groupcommit;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+// An abstract class that has logics and implementations to manage slots and trigger to emit it.
+@ThreadSafe
+abstract class Group<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY, V> {
+  private static final Logger logger = LoggerFactory.getLogger(Group.class);
+
+  protected final Emittable<EMIT_KEY, V> emitter;
+  protected final KeyManipulator<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY> keyManipulator;
+  private final int capacity;
+  private final AtomicReference<Integer> size = new AtomicReference<>();
+  protected final Map<CHILD_KEY, Slot<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY, V>> slots;
+  // Whether to reject a new value slot.
+  protected final AtomicReference<Status> status = new AtomicReference<>(Status.OPEN);
+
+  // Status of the group.
+  enum Status {
+    // Accepting new slot reservation since the number of slots isn't fixed yet.
+    //
+    // Initial status of groups. No group can move back to OPEN.
+    OPEN(false, false, false),
+
+    // Not accepting new slot reservations since the number of slots is already fixed.
+    // Waiting all the slots are set with values.
+    //
+    // Groups with OPEN status can move to CLOSED.
+    CLOSED(true, false, false),
```

Comment:
`CLOSED` corresponds to `OPEN` so it's good, but it sounds like the group has already finished processing.
How about `FIXED`?

Reply:
Thanks for the suggestion! Actually it was first like SIZE-FIXED and I changed it to CLOSED to sound to reject new requests. But as you said, it also sounds things are done. I'll update it to use FIX.

Reply:
Fixed in https://github.com/scalar-labs/scalardb/pull/1607/commits/5000335099a2c15528ef9df89f7ba50e2c4dda82. Used `SIZE_FIXED` to make it clearer.